### PR TITLE
Feature/add examples

### PR DESCRIPTION
**Pull request by ArthurH91**

### CODE REVIEW

**Comment by ArthurH91** [comment]
examples work and everything is alright. Feel free to test it

**Review by Kotochleb — CHANGES_REQUESTED** [review]
Move changes in licensing to another, separate PR.
Meshes can be obtained from a dependency `example-robot-data` and used in the custom URDF file.
Fix JRL modules. This PR is bringing back unpdated version

File: examples/param_parsers.py [hunk 1]
```diff
@@ -0,0 +1,188 @@
...
+
+    def get_target_pose(self):
+        return pin.SE3(
+            pin.Quaternion(
+                *tuple(self.data["TARGET_POSE"]["orientation"])
+            ).toRotationMatrix(),
+            np.array(self.data["TARGET_POSE"]["translation"]),
+        )
+
+    def get_initial_config(self):
```
Comment (Kotochleb) [inline examples/param_parsers.py:139]:
```suggestion
    def get_initial_config(self) -> npt.NDArray[np.float64]:
```

Requires import 
```python
import numpy.typing as npt
```

File: examples/param_parsers.py [hunk 2]
```diff
@@ -0,0 +1,188 @@
...
+        return np.array(self.data["INITIAL_CONFIG"])
+
+    def get_X0(self):
```
Comment (Kotochleb) [inline examples/param_parsers.py:142]:
```suggestion
    def get_X0(self) -> npt.NDArray[np.float64]:
```

File: examples/param_parsers.py [hunk 3]
```diff
@@ -0,0 +1,188 @@
...
+        return np.concatenate(
+            (self.get_initial_config(), np.array(self.data["INITIAL_VELOCITY"]))
+        )
+
+    def get_safety_threshold(self):
```
Comment (Kotochleb) [inline examples/param_parsers.py:147]:
```suggestion
    def get_safety_threshold(self) -> float:
```

File: examples/param_parsers.py [hunk 4]
```diff
@@ -0,0 +1,188 @@
...
+        return self.data["SAFETY_THRESHOLD"]
+
+    def get_T(self):
```
Comment (Kotochleb) [inline examples/param_parsers.py:150]:
```suggestion
    def get_T(self) -> int:
```

File: examples/param_parsers.py [hunk 5]
```diff
@@ -0,0 +1,188 @@
...
+        return self.data["T"]
+
+    def get_dt(self):
```
Comment (Kotochleb) [inline examples/param_parsers.py:153]:
```suggestion
    def get_dt(self) -> float:
```

File: examples/param_parsers.py [hunk 6]
```diff
@@ -0,0 +1,188 @@
...
+        return self.data["dt"]
+
+    def get_di(self):
```
Comment (Kotochleb) [inline examples/param_parsers.py:156]:
```suggestion
    def get_di(self) -> float:
```

File: examples/param_parsers.py [hunk 7]
```diff
@@ -0,0 +1,188 @@
...
+        return self.data["di"]
+
+    def get_ds(self):
```
Comment (Kotochleb) [inline examples/param_parsers.py:159]:
```suggestion
    def get_ds(self) -> float:
```

File: examples/param_parsers.py [hunk 8]
```diff
@@ -0,0 +1,188 @@
...
+        return self.data["ds"]
+
+    def get_ksi(self):
```
Comment (Kotochleb) [inline examples/param_parsers.py:162]:
```suggestion
    def get_ksi(self) -> float:
```

File: examples/param_parsers.py [hunk 9]
```diff
@@ -0,0 +1,188 @@
...
+        return self.data["ksi"]
+
+    def get_W_xREG(self):
```
Comment (Kotochleb) [inline examples/param_parsers.py:165]:
```suggestion
    def get_W_xREG(self) -> float:
```

File: examples/param_parsers.py [hunk 10]
```diff
@@ -0,0 +1,188 @@
...
+        return self.data["WEIGHT_xREG"]
+
+    def get_W_uREG(self):
```
Comment (Kotochleb) [inline examples/param_parsers.py:168]:
```suggestion
    def get_W_uREG(self) -> float:
```

File: examples/param_parsers.py [hunk 11]
```diff
@@ -0,0 +1,188 @@
...
+        return self.data["WEIGHT_uREG"]
+
+    def get_W_gripper_pose(self):
```
Comment (Kotochleb) [inline examples/param_parsers.py:171]:
```suggestion
    def get_W_gripper_pose(self) -> float:
```

File: examples/param_parsers.py [hunk 12]
```diff
@@ -0,0 +1,188 @@
...
+        return self.data["WEIGHT_GRIPPER_POSE"]
+
+    def get_W_gripper_pose_term(self):
```
Comment (Kotochleb) [inline examples/param_parsers.py:174]:
```suggestion
    def get_W_gripper_pose_term(self) -> float:
```

File: examples/param_parsers.py [hunk 13]
```diff
@@ -0,0 +1,188 @@
...
+# All rights reserved.
+
+import hppfcl
+import numpy as np
+import pinocchio as pin
+import yaml
+
+
+class ParamParser:
+    def __init__(self, path: str, scene: int):
```
Comment (Kotochleb) [inline examples/param_parsers.py:14]:
```suggestion
    def __init__(self, path: pathlib.Path, scene: int):
```
Path should always be handled with `pathlib` and not `str`

File: examples/scenes.yaml [hunk 14]
```diff
@@ -0,0 +1,195 @@
...
+
+  # Initial velocity
+  INITIAL_VELOCITY:
+    - 0
+    - 0
+    - 0
+    - 0
+    - 0
+    - 0
+    - 0
```
Comment (Kotochleb) [inline examples/scenes.yaml:26]:
```suggestion
  INITIAL_VELOCITY:
    - 0.0
    - 0.0
    - 0.0
    - 0.0
    - 0.0
    - 0.0
    - 0.0
```

File: examples/scenes.yaml [hunk 15]
```diff
@@ -0,0 +1,195 @@
...
+
+  # Target pose
+  TARGET_POSE:
+    orientation: [ 0, 0, 1, 0]
+    translation: [0, 0.5, 1.2]
```
Comment (Kotochleb) [inline examples/scenes.yaml:31]:
```suggestion
    orientation: [0.0, 0.0, 1.0, 0.0]
    translation: [0.0, 0.5, 1.2]
```

File: examples/scenes.yaml [hunk 16]
```diff
@@ -0,0 +1,195 @@
...
+
+  # Control parameters
+  ksi: 0.0001
+  di: 0.001
+  ds: 0.000001 # 0.0000001
```
Comment (Kotochleb) [inline examples/scenes.yaml:36]:
```suggestion
  ds: 0.000001
```

File: examples/scenes.yaml [hunk 17]
```diff
@@ -0,0 +1,195 @@
...
+
+  # Weights
+  WEIGHT_uREG: 0.0001
+  WEIGHT_xREG: 0.1
+  WEIGHT_GRIPPER_POSE: 1
```
Comment (Kotochleb) [inline examples/scenes.yaml:41]:
```suggestion
  WEIGHT_GRIPPER_POSE: 1.0
```

File: examples/scenes.yaml [hunk 18]
```diff
@@ -0,0 +1,195 @@
...
+  WEIGHT_GRIPPER_POSE_TERM: 50
+  WEIGHT_LIMIT: 0.1
+
+  # Safety threshold
+  SAFETY_THRESHOLD: 0
```
Comment (Kotochleb) [inline examples/scenes.yaml:46]:
```suggestion
  SAFETY_THRESHOLD: 0.0
```

File: examples/scenes.yaml [hunk 19]
```diff
@@ -0,0 +1,195 @@
...
+
+  OBSTACLES:
+    obstacle1:
+        type: "ellipsoid"
+        translation: [0, 0.1, 1.2]
+        orientation: [ 0, 0.707, 0.707, 0]
```
Comment (Kotochleb) [inline examples/scenes.yaml:52]:
```suggestion
        translation: [0.0, 0.1, 1.2]
        orientation: [ 0.0, 0.707, 0.707, 0.0]
```

File: examples/scenes.yaml [hunk 20]
```diff
@@ -0,0 +1,195 @@
...
+    - 1.696128891e-00
+    - 4.77514312e-01
+  INITIAL_VELOCITY:
+    - 0
+    - 0
+    - 0
+    - 0
+    - 0
+    - 0
+    - 0
```
Comment (Kotochleb) [inline examples/scenes.yaml:84]:
```suggestion
  INITIAL_VELOCITY:
    - 0.0
    - 0.0
    - 0.0
    - 0.0
    - 0.0
    - 0.0
    - 0.0
```

File: examples/scenes.yaml [hunk 21]
```diff
@@ -0,0 +1,195 @@
...
+
+  # Target pose
+  TARGET_POSE:
+    orientation: [ 0, 0, 1, 0 ]
+    translation: [0,0.2,0.9]
```
Comment (Kotochleb) [inline examples/scenes.yaml:89]:
```suggestion
    orientation: [0.0, 0.0, 1.0, 0.0 ]
    translation: [0.0, 0.2, 0.9]
```

File: examples/scenes.yaml [hunk 22]
```diff
@@ -0,0 +1,195 @@
...
+  # Control parameters
+  ksi: 0.01
+  di: 0.01
+  ds: 0.0001
+
+  # Weights
+  WEIGHT_uREG: 0.0001
+  WEIGHT_xREG: 0.1
+  WEIGHT_GRIPPER_POSE: 5
+  WEIGHT_GRIPPER_POSE_TERM: 100
```
Comment (Kotochleb) [inline examples/scenes.yaml:100]:
```suggestion
  WEIGHT_GRIPPER_POSE: 5.0
  WEIGHT_GRIPPER_POSE_TERM: 100.0
```

File: examples/scenes.yaml [hunk 23]
```diff
@@ -0,0 +1,195 @@
...
+  WEIGHT_LIMIT: 0.1
+
+  # Safety threshold
+  SAFETY_THRESHOLD: 0.001
+
+  OBSTACLES:
+    obstacle1:
+        type: "ellipsoid"
+        translation: [0.0, 0.0, 0.9]
+        orientation: [0.707,0, 0.707, 0]
```
Comment (Kotochleb) [inline examples/scenes.yaml:110]:
```suggestion
        orientation: [0.707, 0.0, 0.707, 0.0]
```

File: examples/scenes.yaml [hunk 24]
```diff
@@ -0,0 +1,195 @@
...
+
+# Initial velocity (all zeros)
+  INITIAL_VELOCITY:
+    - 0
+    - 0
+    - 0
+    - 0
+    - 0
+    - 0
+    - 0
```
Comment (Kotochleb) [inline examples/scenes.yaml:159]:
```suggestion
  INITIAL_VELOCITY:
    - 0.0
    - 0.0
    - 0.0
    - 0.0
    - 0.0
    - 0.0
    - 0.0
```

File: examples/scenes.yaml [hunk 25]
```diff
@@ -0,0 +1,195 @@
...
+
+  # Initial state vector
+  x0: [-0.566 ,0.8085,-1.0189,-0.4379,0.7698,1.092, -0.2002, 0, 0, 0, 0, 0, 0, 0]
```
Comment (Kotochleb) [inline examples/scenes.yaml:162]:
```suggestion
  x0: [-0.566, 0.8085, -1.0189, -0.4379, 0.7698, 1.092, -0.2002, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0]
```

File: examples/scenes.yaml [hunk 26]
```diff
@@ -0,0 +1,195 @@
...
+
+  DIM_OBS:
+  - [1.0, 0.25, 0.1]
+
+  # Target pose
+  TARGET_POSE:
+    orientation: [ 0, 0, 1, 0 ]
```
Comment (Kotochleb) [inline examples/scenes.yaml:169]:
```suggestion
    orientation: [0.0, 0.0, 1.0, 0.0]
```

File: examples/scenes.yaml [hunk 27]
```diff
@@ -0,0 +1,195 @@
...
+  # Control parameters
+  ksi: 0.001
+  di: 0.01  # 1e-4
+  ds: 0.0001
+
+  # Weights
+  WEIGHT_uREG: 0.0001
+  WEIGHT_xREG: 0.1
+  WEIGHT_GRIPPER_POSE: 5
+  WEIGHT_GRIPPER_POSE_TERM: 100
```
Comment (Kotochleb) [inline examples/scenes.yaml:181]:
```suggestion
  WEIGHT_GRIPPER_POSE: 5.0
  WEIGHT_GRIPPER_POSE_TERM: 100.0
```

File: examples/scenes.yaml [hunk 28]
```diff
@@ -0,0 +1,195 @@
...
+  WEIGHT_LIMIT: 0.1
+
+  # Safety threshold
+  SAFETY_THRESHOLD: 0
```
Comment (Kotochleb) [inline examples/scenes.yaml:185]:
```suggestion
  SAFETY_THRESHOLD: 0.0
```

File: examples/scenes.yaml [hunk 29]
```diff
@@ -0,0 +1,195 @@
...
+
+  OBSTACLES:
+    obstacle1:
+        type: "ellipsoid"
+        translation: [0.1, 0.1, 1.2]
+        orientation: [0.7071,0.7071, 0, 0]
```
Comment (Kotochleb) [inline examples/scenes.yaml:191]:
```suggestion
        orientation: [0.7071, 0.7071, 0.0, 0.0]
```

File: examples/visualizer.py [hunk 30]
```diff
@@ -0,0 +1,92 @@
...
+GREEN_FULL = np.array([170, 236, 149, 255]) / 255
+
+BLUE = np.array([144, 169, 183, 125]) / 255
+BLUE_FULL = np.array([144, 169, 183, 255]) / 255
+
+YELLOW = np.array([1, 1, 0, 0.5])
+YELLOW_FULL = np.array([1, 1, 0, 1.0])
+
+BLACK = np.array([0, 0, 0, 0.5])
+BLACK_FULL = np.array([0, 0, 0, 1.0])
```
Comment (Kotochleb) [inline examples/visualizer.py:24]:
```suggestion
RED = np.array([249, 136, 126, 125]) / 255.0
RED_FULL = np.array([249, 136, 126, 255]) / 255.0

GREEN = np.array([170, 236, 149, 125]) / 255.0
GREEN_FULL = np.array([170, 236, 149, 255]) / 255.0

BLUE = np.array([144, 169, 183, 125]) / 255.0
BLUE_FULL = np.array([144, 169, 183, 255]) / 255.0

YELLOW = np.array([1.0, 1.0, 0.0, 0.5])
YELLOW_FULL = np.array([1.0, 1.0, 0.0, 1.0])

BLACK = np.array([0.0, 0.0, 0.0, 0.5])
BLACK_FULL = np.array([0.0, 0.0, 0.0, 1.0])
```

File: examples/visualizer.py [hunk 31]
```diff
@@ -0,0 +1,92 @@
...
+        model=rmodel,
+        collision_model=cmodel,
+        visual_model=vmodel,
+    )
+    viz.initViewer(viewer=meshcat.Visualizer(zmq_url="tcp://127.0.0.1:6000"))
+    viz.clean()
+    viz.loadViewerModel("pinocchio")
+
+    viz.displayCollisions(True)
+    return viz
```
Comment (Kotochleb) [inline examples/visualizer.py:38]:
```suggestion
def create_viewer(
    rmodel: pin.Model, cmodel: pin.GeometryModel, vmodel: pin.GeometryModel
) -> visualize.MeshcatVisualizer:
    viz = visualize.MeshcatVisualizer(
        model=rmodel,
        collision_model=cmodel,
        visual_model=vmodel,
    )
    viz.initViewer(viewer=meshcat.Visualizer(zmq_url="tcp://127.0.0.1:6002"))
    viz.clean()
    viz.loadViewerModel("pinocchio")

    viz.displayCollisions(True)
    return viz
```

requires:
```python
import pinocchio as pin
```

File: examples/visualizer.py [hunk 32]
```diff
@@ -0,0 +1,92 @@
...
+BLUE_FULL = np.array([144, 169, 183, 255]) / 255
+
+YELLOW = np.array([1, 1, 0, 0.5])
+YELLOW_FULL = np.array([1, 1, 0, 1.0])
+
+BLACK = np.array([0, 0, 0, 0.5])
+BLACK_FULL = np.array([0, 0, 0, 1.0])
+
+
+def create_viewer(rmodel, cmodel, vmodel):
```
Comment (Kotochleb) [inline examples/visualizer.py:27]: Missing docstring

File: examples/param_parsers.py [hunk 33]
```diff
@@ -0,0 +1,188 @@
...
+    def get_initial_config(self):
+        return np.array(self.data["INITIAL_CONFIG"])
+
+    def get_X0(self):
+        return np.concatenate(
+            (self.get_initial_config(), np.array(self.data["INITIAL_VELOCITY"]))
+        )
+
+    def get_safety_threshold(self):
+        return self.data["SAFETY_THRESHOLD"]
```
Comment (Kotochleb) [inline examples/param_parsers.py:148]:
```suggestion
        return float(self.data["SAFETY_THRESHOLD"])
```

File: examples/param_parsers.py [hunk 34]
```diff
@@ -0,0 +1,188 @@
...
+
+    def get_T(self):
+        return self.data["T"]
```
Comment (Kotochleb) [inline examples/param_parsers.py:151]:
```suggestion
        return int(self.data["T"])
```

File: examples/param_parsers.py [hunk 35]
```diff
@@ -0,0 +1,188 @@
...
+
+    def get_dt(self):
+        return self.data["dt"]
```
Comment (Kotochleb) [inline examples/param_parsers.py:154]:
```suggestion
        return float(self.data["dt"])
```

File: examples/param_parsers.py [hunk 36]
```diff
@@ -0,0 +1,188 @@
...
+
+    def get_di(self):
+        return self.data["di"]
```
Comment (Kotochleb) [inline examples/param_parsers.py:157]:
```suggestion
        return float(self.data["di"])
```

File: examples/param_parsers.py [hunk 37]
```diff
@@ -0,0 +1,188 @@
...
+
+    def get_ds(self):
+        return self.data["ds"]
```
Comment (Kotochleb) [inline examples/param_parsers.py:160]:
```suggestion
        return float(self.data["ds"])
```

File: examples/param_parsers.py [hunk 38]
```diff
@@ -0,0 +1,188 @@
...
+
+    def get_ksi(self):
+        return self.data["ksi"]
```
Comment (Kotochleb) [inline examples/param_parsers.py:163]:
```suggestion
        return float(self.data["ksi"])
```

File: examples/param_parsers.py [hunk 39]
```diff
@@ -0,0 +1,188 @@
...
+
+    def get_W_xREG(self):
+        return self.data["WEIGHT_xREG"]
```
Comment (Kotochleb) [inline examples/param_parsers.py:166]:
```suggestion
        return float(self.data["WEIGHT_xREG"])
```

File: examples/param_parsers.py [hunk 40]
```diff
@@ -0,0 +1,188 @@
...
+
+    def get_W_uREG(self):
+        return self.data["WEIGHT_uREG"]
```
Comment (Kotochleb) [inline examples/param_parsers.py:169]:
```suggestion
        return float(self.data["WEIGHT_uREG"])
```

File: examples/param_parsers.py [hunk 41]
```diff
@@ -0,0 +1,188 @@
...
+
+    def get_W_gripper_pose(self):
+        return self.data["WEIGHT_GRIPPER_POSE"]
```
Comment (Kotochleb) [inline examples/param_parsers.py:172]:
```suggestion
        return float(self.data["WEIGHT_GRIPPER_POSE"])
```

File: examples/param_parsers.py [hunk 42]
```diff
@@ -0,0 +1,188 @@
...
+
+    def get_W_gripper_pose_term(self):
+        return self.data["WEIGHT_GRIPPER_POSE_TERM"]
```
Comment (Kotochleb) [inline examples/param_parsers.py:175]:
```suggestion
        return float(self.data["WEIGHT_GRIPPER_POSE_TERM"])
```

File: examples/visualizer.py [hunk 43]
```diff
@@ -0,0 +1,92 @@
...
+    )
+    viz.initViewer(viewer=meshcat.Visualizer(zmq_url="tcp://127.0.0.1:6000"))
+    viz.clean()
+    viz.loadViewerModel("pinocchio")
+
+    viz.displayCollisions(True)
+    return viz
+
+
+def add_sphere_to_viewer(viz, sphere_name, radius, position, color=int):
```
Comment (Kotochleb) [inline examples/visualizer.py:41]:
```suggestion
def add_sphere_to_viewer(
    viz: visualize.MeshcatVisualizer,
    sphere_name: str,
    radius: float,
    position: Union[list, npt.NDArray[np.float64]],
    color: int,
) -> None:
```

Requires:
```python
from typing import Union
```

File: examples/visualizer.py [hunk 44]
```diff
@@ -0,0 +1,92 @@
...
+    Parameters:
+    viz (MeshcatVisualizer): The Meshcat visualizer.
+    sphere_name (str): The name of the sphere.
+    radius (float): The radius of the sphere.
+    position (list or np.array): The position of the sphere as [x, y, z].
+    color (int): The color of the sphere as ?.
+    """
+    try:
+        # Check if the sphere already exists
+        _ = viz.viewer[sphere_name]
```
Comment (Kotochleb) [inline examples/visualizer.py:54]:
```suggestion
        viz.viewer[sphere_name]
```

File: examples/visualizer.py [hunk 45]
```diff
@@ -0,0 +1,92 @@
...
+    sphere_geom = meshcat.geometry.Sphere(radius)
+    sphere_material = meshcat.geometry.MeshLambertMaterial(color=color)
+
+    viz.viewer[sphere_name].set_object(sphere_geom, sphere_material)
+    viz.viewer[sphere_name].set_transform(
+        meshcat.transformations.translation_matrix(position)
+    )
+
+
+def add_cube_to_viewer(viz, cube_name, dim, position, color):
```
Comment (Kotochleb) [inline examples/visualizer.py:68]:
```suggestion
def add_cube_to_viewer(
    viz: visualize.MeshcatVisualizer,
    cube_name: str,
    dim: Union[list, npt.NDArray[np.float64]],
    position: Union[list, npt.NDArray[np.float64]],
    color: int,
) -> None:
```

File: examples/visualizer.py [hunk 46]
```diff
@@ -0,0 +1,92 @@
...
+    Parameters:
+    viz (MeshcatVisualizer): The Meshcat visualizer.
+    cube_name (str): The name of the sphere.
+    radius (float): The radius of the sphere.
+    position (list or np.array): The position of the sphere as [x, y, z].
+    color (int): The color of the sphere as ?.
+    """
+    try:
+        # Check if the sphere already exists
+        _ = viz.viewer[cube_name]
```
Comment (Kotochleb) [inline examples/visualizer.py:81]:
```suggestion
        viz.viewer[cube_name]
```

File: examples/wrapper_panda.py [hunk 47]
```diff
@@ -0,0 +1,143 @@
...
+
+RED = np.array([249, 136, 126, 125]) / 255
+
+
+class PandaWrapper:
+    def __init__(
+        self,
+        auto_col=False,
+        capsule=False,
+    ):
```
Comment (Kotochleb) [inline examples/wrapper_panda.py:23]:
```suggestion
    def __init__(
        self,
        auto_col: bool = False,
        capsule: bool = False,
    ) -> None:
```

File: examples/wrapper_panda.py [hunk 48]
```diff
@@ -0,0 +1,143 @@
...
+from os.path import abspath, dirname, join
+
+import hppfcl
+import numpy as np
+import pinocchio as pin
+
+# This class is for unwrapping an URDF and converting it to a model. It is also possible to add objects in the model,
+# such as a ball at a specific position.
+
+RED = np.array([249, 136, 126, 125]) / 255
```
Comment (Kotochleb) [inline examples/wrapper_panda.py:15]:
```suggestion
RED = np.array([249, 136, 126, 125]) / 255.0
```

File: examples/wrapper_panda.py [hunk 49]
```diff
@@ -0,0 +1,143 @@
+# BSD 3-Clause License
+#
+# Copyright (C) 2024, LAAS-CNRS.
+# Copyright note valid unless otherwise stated in individual files.
+# All rights reserved.
+from os.path import abspath, dirname, join
```
Comment (Kotochleb) [inline examples/wrapper_panda.py:6]: Use `pathlib`

File: examples/wrapper_panda.py [hunk 50]
```diff
@@ -0,0 +1,143 @@
...
+        pinocchio_model_dir = dirname(str(abspath(__file__)))
+        model_path = join(pinocchio_model_dir, "models")
+        self._mesh_dir = join(model_path, "meshes")
+        urdf_filename = "franka2.urdf"
+        srdf_filename = "demo.srdf"
+        self._urdf_model_path = join(join(model_path, "urdf"), urdf_filename)
+        self._srdf_model_path = join(join(model_path, "srdf"), srdf_filename)
+
+        # Color of the robot
+        self._color = np.array([249, 136, 126, 255]) / 255
```
Comment (Kotochleb) [inline examples/wrapper_panda.py:41]:
```suggestion
        self._color = np.array([249, 136, 126, 255]) / 255.0
```

File: examples/wrapper_panda.py [hunk 51]
```diff
@@ -0,0 +1,143 @@
...
+
+        # Boolean describing whether the auto-collisions are in the collision model or not
+        self._auto_col = auto_col
+
+        # Transforming the robot from cylinders/spheres to capsules
+        self._capsule = capsule
+
+    def __call__(self):
```
Comment (Kotochleb) [inline examples/wrapper_panda.py:49]:
```suggestion
    def __call__(self) -> Tuple[pin.Model, pin.GeometryModel, pin.GeometryModel]:
```

Requires:
```python
from typing import Tuple
````

File: examples/wrapper_panda.py [hunk 52]
```diff
@@ -0,0 +1,143 @@
...
+            self._model_reduced, rdata, self._cmodel_reduced, cdata, q0
+        )
+
+        return (
+            self._model_reduced,
+            self._cmodel_reduced,
+            self._vmodel_reduced,
+        )
+
+    def transform_model_into_capsules(self):
```
Comment (Kotochleb) [inline examples/wrapper_panda.py:110]:
```suggestion
    def transform_model_into_capsules(self) -> None:
```

File: examples/wrapper_panda.py [hunk 53]
```diff
@@ -0,0 +1,143 @@
...
+
+        # Going through all the goemetry objects in the collision model
+        for geom_object in collision_model_reduced_copy.geometryObjects:
+            if isinstance(geom_object.geometry, hppfcl.Cylinder):
+                # Sometimes for one joint there are two cylinders, which need to be defined by two capsules for the same link.
+                # Hence the name convention here.
+                if (geom_object.name[:-4] + "capsule_0") in list_names_capsules:
+                    name = geom_object.name[:-4] + "capsule_" + "1"
+                else:
+                    name = geom_object.name[:-4] + "capsule_" + "0"
```
Comment (Kotochleb) [inline examples/wrapper_panda.py:123]:
```suggestion
                if (geom_object.name[:-4] + "capsule_0") in list_names_capsules:
                    name = geom_object.name[:-4] + "capsule_1"
                else:
                    name = geom_object.name[:-4] + "capsule_0"
```

File: examples/wrapper_panda.py [hunk 54]
```diff
@@ -0,0 +1,143 @@
...
+    def transform_model_into_capsules(self):
+        """Modifying the collision model to transform the spheres/cylinders into capsules which makes it easier to have a fully constrained robot."""
+        collision_model_reduced_copy = self._cmodel_reduced.copy()
+        list_names_capsules = []
+
+        # Going through all the goemetry objects in the collision model
+        for geom_object in collision_model_reduced_copy.geometryObjects:
+            if isinstance(geom_object.geometry, hppfcl.Cylinder):
+                # Sometimes for one joint there are two cylinders, which need to be defined by two capsules for the same link.
+                # Hence the name convention here.
```
Comment (Kotochleb) [inline examples/wrapper_panda.py:119]:
```suggestion
                # Sometimes for one joint there are two cylinders,
                # which need to be defined by two capsules for the same link.
                # Hence the name convention here.
```

File: examples/main_ocp.py [hunk 55]
```diff
@@ -0,0 +1,108 @@
...
+ocp_vel.solve(XS_init, US_init, 100)
+
+
+print("OCP with velocity constraints solved")
+for i, xs in enumerate(ocp_vel.xs):
+    q = np.array(xs[:7].tolist())
+    pin.framesForwardKinematics(rmodel, rdata, q)
+    add_cube_to_viewer(
+        vis,
+        "vcolmpc" + str(i),
```
Comment (Kotochleb) [inline examples/main_ocp.py:83]:
```suggestion
        f"vcolmpc{i}",,
```

File: examples/main_ocp.py [hunk 56]
```diff
@@ -0,0 +1,108 @@
...
+        rdata.oMf[rmodel.getFrameId("panda2_rightfinger")].translation,
+        color=100000000,
+    )
+
+for i, xs in enumerate(OCP_dist.xs):
+    q = np.array(xs[:7].tolist())
+    pin.framesForwardKinematics(rmodel, rdata, q)
+    add_sphere_to_viewer(
+        vis,
+        "colmpc" + str(i),
```
Comment (Kotochleb) [inline examples/main_ocp.py:94]:
```suggestion
        f"colmpc{i}",
```

File: examples/main_ocp.py [hunk 57]
```diff
@@ -0,0 +1,108 @@
...
+        2e-2,
+        rdata.oMf[rmodel.getFrameId("panda2_rightfinger")].translation,
+        color=100000,
+    )
+print("Press enter to display the solution")
```
Comment (Kotochleb) [inline examples/main_ocp.py:99]:
```suggestion
print("Press 'ENTER' to display the solution")
```

File: README.md [hunk 58]
```diff
@@ -31,16 +29,13 @@ HPP-FCL & Pinocchio must be built from sources. Build pinocchio with the flag :
...
+- If you have a problem with ```FakeCollisionGeometry```, it is likely that the linking of Pinocchio with HPPFCL wasn't done properly. Verify that you have the right commits & the right compilation flags.
+- The main branch of HPPFCL doesn't compute well the closest points and thus, this repo needs to be built upon the devel branch. If it built but doesn't avoid collision, make sure that you didn't built the main branch.
 
-### For the MPC part:
+### For the OCP part:
 
-Create a meshcat-server by writting in a terminal ```meshcat-server```. Then, to see the example with hard constraints simply run in the main directory ```python examples/mpc_panda_reaching.py```. To compare it with the standard MPC without hard constraint but collision avoidance written as soft constraint, run  ```python examples/mpc_panda_reaching_soft_constraint.py```.
+Create a meshcat-server by writting in a terminal ```meshcat-server```. Then, to see the example with hard constraints simply run in the main directory ```python examples/main_ocp.py -s i```, where i is the index of the scenario, going from 1 to 3.
```
Comment (Kotochleb) [inline README.md:38]:
```suggestion
Create a meshcat-server by writting in a terminal ```meshcat-server --zmq-url "tcp://127.0.0.1:6002"```. Then, to see the example with hard constraints simply run in the main directory ```python examples/main_ocp.py -s i```, where i is the index of the scenario, going from 1 to 3.
```
Starting `meshcat-server` externally not always guarantees that the port will be 6000, sometimes it starts at a different port and then needs to be manually adjusted. At least in my case when I wanted to start it with port 6000 I not always could do so

**Review by Kotochleb — CHANGES_REQUESTED** [review]
Some very minor suggestions. Otherwise the code looks good

File: README.md (hunk 58 above)
Comment (ArthurH91) [reply README.md:38]: let's have a meshcat server run every time the example is ran

**Review by Kotochleb — APPROVED** [review]
Look ok

**Pull request merged**